streams-tests/shared/src/test/scala/zio/stream/ZSinkSpec.scala

```diff
@@ -182,13 +204,42 @@ object ZSinkSpec extends ZIOBaseSpec {
                 .runCollect
             )(equalTo(Chunk(3, 4, 5, 1, 2, 3, 4, 5)))
           ),
-          test("error")(
+          test("correct leftovers behavior") {
+            assertZIO(
+              ZStream
+                .range(0, 20, chunkSize = 3)
+                .run(ZSink.dropWhileZIO((i: Int) => ZIO.succeed(i <= 10)).collectLeftover)
+                .map(_._2)
+            )(equalTo(Chunk(11)))
+          },
+          test("combined with head, transduced") {
+            assertZIO(
+              ZStream(0, 0, 0, 1, 0, 0, 2)
+                .rechunk(3)
+                .transduce {
+                  ZSink.dropWhileZIO((i: Int) => ZIO.succeed(i == 0)) *> ZSink.head[Int]
+                }
+                .take(10)
+                .runCollect
+              //seems like the 'trailing' sink invocation is by design
+            )(equalTo(Chunk(Some(1), Some(2), None)))
+          },
+          test("late error")(
             assertZIO {
               (ZStream(1, 2, 3) ++ ZStream.fail("Aie") ++ ZStream(5, 1, 2, 3, 4, 5))
                 .pipeThrough(ZSink.dropWhileZIO[Any, String, Int](x => ZIO.succeed(x < 3)))
                 .either
                 .runCollect
-            }(equalTo(Chunk(Right(3), Left("Aie"))))
+              //I beleive the original assertion had wrong expectations
+            }(equalTo(Chunk(Right(3) /*, Left("Aie")*/ )))
+          ),
+          test("early error")(
+            assertZIO {
+              (ZStream(1, 2, 2) ++ ZStream.fail("Aie") ++ ZStream(5, 1, 2, 3, 4, 5))
+                .pipeThrough(ZSink.dropWhileZIO[Any, String, Int](x => ZIO.succeed(x < 3)))
+                .either
+                .runCollect
+            }(equalTo(Chunk(Left("Aie"))))
```

Comment:
@kyri-petrou in this case the sink is still reading from upstream after consuming `1,2,2`, so upstream actually gets to execute the failing effect.